source/opt/optimizer.cpp

```diff
@@ -66,19 +66,50 @@ Optimizer& Optimizer::RegisterPass(PassToken&& p) {
+// the code will be invalid at the start:
+//
+// 1) There will of opaque objects, like images, which will be passed around
+//    in intermediate objects.  Valid spir-v will have to replace the use of
+//    the opaque object with an intermediate object that is the result of the
+//    load of the opaque object.
+//
+// 2) There will be variables that contain pointers to structured or uniform
+//    buffers.  It be legal, the variables must be eliminated, and the
+//    references to the structured buffers must use the result of OpVariable
+//    in the Uniform storage class.
+//
+// Optimization in this list must accept shaders with these relaxation of the
+// rules.  There is not guarantee that this list of optimizations is able to
+// legalize all inputs, but it is on a best effort basis.
+//
+// The legalization problem is essentially a very general copy propagation
+// problem.  The optimization we use are all used to either do copy propagation
+// or enable more copy propagation.
 Optimizer& Optimizer::RegisterLegalizationPasses() {
-  return RegisterPass(CreateInlineExhaustivePass())
-      .RegisterPass(CreateLocalAccessChainConvertPass())
-      .RegisterPass(CreateLocalSingleBlockLoadStoreElimPass())
-      .RegisterPass(CreateLocalSingleStoreElimPass())
-      .RegisterPass(CreateInsertExtractElimPass())
-      .RegisterPass(CreateAggressiveDCEPass())
-      .RegisterPass(CreateDeadBranchElimPass())
-      .RegisterPass(CreateCFGCleanupPass())
-      .RegisterPass(CreateBlockMergePass())
-      .RegisterPass(CreateLocalMultiStoreElimPass())
-      .RegisterPass(CreateInsertExtractElimPass())
-      .RegisterPass(CreateAggressiveDCEPass());
+  return
+      // Make sure uses and definitions are in the same function.
+      RegisterPass(CreateInlineExhaustivePass())
+          // Make private variable function scope
+          .RegisterPass(CreateEliminateDeadFunctionsPass())
+          .RegisterPass(CreatePrivateToLocalPass())
+          // Split up aggragates so they are easier to deal with.
+          .RegisterPass(CreateScalarReplacementPass())
+          // Remove loads and stores so everything is in intermediate values.
+          // Takes care of copy propagation of non-members.
+          .RegisterPass(CreateLocalMultiStoreElimPass())
+          // Copy propagate members.  Cleans up code sequences generated by
+          // scalar replacement.
+          .RegisterPass(CreateInsertExtractElimPass())
+          // TODO: Add constant propagation here
+          // May need loop unrolling here see
+          // https://github.com/Microsoft/DirectXShaderCompiler/pull/930
+          .RegisterPass(CreateCFGCleanupPass())
+          // Get rid of unused code that leave traces of the illegal code.
+          .RegisterPass(CreateAggressiveDCEPass())
+          // TODO: Remove this once ADCE can do it.
+          .RegisterPass(CreateDeadVariableEliminationPass());
```

Comment:
Can we have the auxiliary `CreateCompactIdsPass()` here? Running this recipe bloats the result-id ranges used.

Reply:
But ID compaction is not strictly a legalization.  If you think it's beneficial for other reasons then you should apply it separately. ??

Reply:
That's true. I can register it separately then.